From: Stefan Haberland <sth@linux.vnet.ibm.com>
Subject: s390/dasd: suppress command reject error for query host access command
Patch-mainline: v4.12-rc1
Git-commit: ab24fbd35a6ee77a58c24bd50582c51610a194f0
References: bnc#1038912, LTC#153591

Description:  dasd: Prevent error message when using query host access.
Symptom:      An error message occurs when using the query host access
              feature on storage servers that do not support it.
              This is implicitly done when using DASD tools like dasdfmt
              or fdasd on such a device.
Problem:      The error message is not clearly related to the query host
              access feature and looks like a generic DASD I/O error which
              might confuse customers.
Solution:     Check if the query host access feature is available on the
              storage server to not use it. This might not be
              sufficient for all cases. Suppress the error message for all
              remaining cases.
Reproduction: Read the host_access_count sysfs attribute for a storage
              server that does not support it.

Upstream-Description:

              s390/dasd: suppress command reject error for query host access command

              On some z/VM systems the query host access command is not supported for
              temp disks, though the corresponding feature code is set.
              This does not have any impact beside that the information is not available.
              Suppress the full blown command reject error messages to not confuse the
              user. The error is still logged in the s390dbf.

              Signed-off-by: Stefan Haberland <sth@linux.vnet.ibm.com>
              Signed-off-by: Martin Schwidefsky <schwidefsky@de.ibm.com>


Signed-off-by: Stefan Haberland <sth@linux.vnet.ibm.com>
Acked-by: John Jolly <jjolly@suse.com>
---
 drivers/s390/block/dasd_3990_erp.c |    5 +++--
 drivers/s390/block/dasd_eckd.c     |   12 +++++++++---
 drivers/s390/block/dasd_int.h      |    2 +-
 3 files changed, 13 insertions(+), 6 deletions(-)

--- a/drivers/s390/block/dasd_3990_erp.c
+++ b/drivers/s390/block/dasd_3990_erp.c
@@ -1052,8 +1052,9 @@ dasd_3990_erp_com_rej(struct dasd_ccw_re
 	} else {
 		/* fatal error -  set status to FAILED
 		   internal error 09 - Command Reject */
-		dev_err(&device->cdev->dev, "An error occurred in the DASD "
-			"device driver, reason=%s\n", "09");
+		if (!test_bit(DASD_CQR_SUPPRESS_CR, &erp->flags))
+			dev_err(&device->cdev->dev,
+				"An error occurred in the DASD device driver, reason=09\n");
 
 		erp = dasd_3990_erp_cleanup(erp, DASD_CQR_FAILED);
 	}
--- a/drivers/s390/block/dasd_eckd.c
+++ b/drivers/s390/block/dasd_eckd.c
@@ -4914,10 +4914,14 @@ static void dasd_eckd_dump_sense(struct
 		dasd_eckd_dump_sense_tcw(device, req, irb);
 	} else {
 		/*
-		 * In some cases the 'No Record Found' error might be expected
-		 * and log messages shouldn't be written then. Check if the
-		 * according suppress bit is set.
+		 * In some cases the 'Command Reject' or 'No Record Found'
+		 * error might be expected and log messages shouldn't be
+		 * written then. Check if the according suppress bit is set.
 		 */
+		if (sense && sense[0] & SNS0_CMD_REJECT &&
+		    test_bit(DASD_CQR_SUPPRESS_CR, &req->flags))
+			return;
+
 		if (sense && sense[1] & SNS1_NO_REC_FOUND &&
 		    test_bit(DASD_CQR_SUPPRESS_NRF, &req->flags))
 			return;
@@ -5206,6 +5210,8 @@ static int dasd_eckd_query_host_access(s
 
 	cqr->buildclk = get_tod_clock();
 	cqr->status = DASD_CQR_FILLED;
+	/* the command might not be supported, suppress error message */
+	__set_bit(DASD_CQR_SUPPRESS_CR, &cqr->flags);
 	rc = dasd_sleep_on_interruptible(cqr);
 	if (rc == 0) {
 		*data = *host_access;
--- a/drivers/s390/block/dasd_int.h
+++ b/drivers/s390/block/dasd_int.h
@@ -239,11 +239,11 @@ struct dasd_ccw_req {
 					 */
 /*
  * The following flags are used to suppress output of certain errors.
- * These flags should only be used for format checks!
  */
 #define DASD_CQR_SUPPRESS_NRF	4	/* Suppress 'No Record Found' error */
 #define DASD_CQR_SUPPRESS_FP	5	/* Suppress 'File Protected' error*/
 #define DASD_CQR_SUPPRESS_IL	6	/* Suppress 'Incorrect Length' error */
+#define DASD_CQR_SUPPRESS_CR	7	/* Suppress 'Command Reject' error */
 
 /* Signature for error recovery functions. */
 typedef struct dasd_ccw_req *(*dasd_erp_fn_t) (struct dasd_ccw_req *);
